### PR TITLE
Android: Enable sharing to Joplin on Android

### PR DESCRIPTION
**Pull request by roman-r-m**
This is based on react-native-share-extension. Not sure what was the issue that broke that functionality in the first place. react-native-share-extension seems abandoned and I do not want to fork and maintain it so I reimplemented the required parts (and some more) here.

A few more things to do (here or as separate items):
- [x] Share images
- [x] Share documents (e.g. pdf)
- [x] Improve UI: atm sharing uses the same UI as normal editing, might be better to have a modal window (something like Google Keep).

I'd appreciate if someone could try this on iOS just to make sure I didn't break anything.

Closes https://github.com/laurent22/joplin/issues/2896

### CODE REVIEW

File: ReactNativeClient/lib/share.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+const { NativeModules, Platform } = require('react-native');
+
+const ext = Platform.OS === 'android' ?
```
Comment (roman-r-m) [inline ReactNativeClient/lib/share.js]: Tried to make sure it does not break iOS but can't test myself as I do not have any iOS devices

**Comment by CalebJohn** [comment]
Hi, thanks for starting on this a few comments.

First of all, it seems like the bulk of the line changes in this pull come from whitespace updates. This makes code review very difficult as we can not focus in on whats actually changed. So please undo the purely whitespace changes you've made (it seems like your editor has caused this).
The second thing, what exactly have you changed here? It looks like you've simply uncommented some code? Am I missing something?

**Comment by roman-r-m** [comment]
Hi CalebJohn

Thanks for looking into this. The whitespace changes indeed come from IDEA, I'll revert them.

As for what exactly have been changed, indeed it's mostly uncommenting and fixing a few bits and pieces like adding title.
I've seen on the forum that this feature had stopped working after react native upgrade but I did not check why. I started implementing it from scratch but ended up reusing most of the code that was already there.

**Comment by tessus** [comment]
There's actually a GSoC project/idea about sharing. While I do find it great that this PR restores the previous functionality, sharing needs definitely more attention.
It only worked for text and only on Android.

So the new sharing solution must support:

- Android / iOS
- text, images, and attachments (binary files)

(I would like to see this functionality on desktop as well, but that's another story...)

**Comment by roman-r-m** [comment]
Hi

I do plan to add support for images and other data types but only for
android. I do not have any iOS devices or any experience developing for
iOS. Plus my understanding was it only worked on Android anyway.

**Comment by tessus** [comment]
I just want to make sure that your efforts are not for nothing.

e.g. in about a week, we'll decide, if we enroll someone for the [sharing idea](https://joplinapp.org/gsoc2020/ideas.html#4-sharing-on-mobile) for GSoC. If you now start to code something specifically for Android, it could be that your code will be backed out.

None of this is my decision, I just wanted to make you aware of it.

**Comment by roman-r-m** [comment]
I did it as it's a feature I need and I didn't want to wait for it be eventually implemented.

If what I wrote would be useful when coding for GSOC -- good, if not at least I learned some Android & React Native (I have almost zero experience in either of them).

**Comment by roman-r-m** [comment]
@tessus has the decision been made on this feature and GSoC? I got it to the point where all functionality I need (share any number of items of any type) is working.
It's a bit rought aroun edges but for my purposes it's enough. So the question now is should I make it PR-ready?

**Comment by tessus** [comment]
The deadline for the proposals is today, so after that we have 2 or 3 weeks to decide which projects and students we'll go with.

**Comment by roman-r-m** [comment]
I see, thank you. I thought it was going to be sooner (you previously wrote "e.g. in about a week, we'll decide").

**Comment by tessus** [comment]
Yea, sorry about that. I should have expressed myself more clearly. I actually meant that we'll begin our decision process. My bad.

**Comment by tessus** [comment]
With regards to a PR: I'm not quite sure. @laurent22 , @PackElend what do you guys think?

**Comment by PackElend** [comment]
we have 4 proposal for sharing let us have quick look and feedback here

**Comment by PackElend** [comment]
@roman-r-m you can proceed with the PR

**Comment by roman-r-m** [comment]
@PackElend thanks for confirming.

**I updated the PR but it's still not quite done yet.**

There are a few issues:
- [ ] If I share something then go back to Joplin, open a note and then go back, the app closes, seems fromShare flag is not reset here https://github.com/laurent22/joplin/pull/2870/files#diff-b185419b435222220c2d114eb5aba85dR308
- [ ] Getting occasional errors "Can't perform a React state update..." -- still have not found the root cause
- [ ] Need to test how it would work with encrypted notes and make sure it won't allow access to notes simply by sharing to Joplin

**Comment by roman-r-m** [comment]
@laurent22 @CalebJohn I could use some advice about an issue I'm seeing when debugging this PR.

The scenario is:
- open a note
- switch to another app
- share anything to Joplin

At this point I see this warning when running in debugger:
```
Warning: Can't perform a React state update on an unmounted component. This is a no-op, but it indicates a memory leak in your application. To fix, cancel all subscriptions and asynchronous tasks in the componentWillUnmount method.
    in NoteScreenComponent (created by Connect(NoteScreenComponent))
    in Connect(NoteScreenComponent) (at app-nav.js:74)
```

I have added debugging statemens and noticed that when I dispatch NAV_BACK at https://github.com/laurent22/joplin/pull/2870/files#diff-61c345cdd894731a1b716b473ead6decR672 following the pattern that is used elsewhere in the app, I could see NoteScreenComponent being unmounted twice which I think is causing the above issue.

What, I think, is happening is Android starts a new activity (ShareActivity). This actvity is mapped to the same react component via the ShareActivity.getMainComponentName() method. React creates a new instance of the App component that shares the Redux store with the original instance that corresponds to Joplin itself (and MainActivity from the OS perspective) and since the store is shared both components will receive NAV_BACK commmand or any other command for that matter.

Does that sounds plausible?

I can see 2 ways to solve this:
* Keep using the single Redux store (I believe this is generally recommended when using Redux?) but somehow indicate which component to dispatch to, something like this:
```
this.props.dispatch({ type: 'NAV_BACK', target: 'ShareApp' });
```

* Create a separate Redux store per component. I took a stab at this but had some other issues which I've not managed to resolve yet.

I personally think the 2nd approach is better but would like your input on this.

Thanks.

**Comment by roman-r-m** [comment]
This PR is now ready for review. I had to put a few adhoc fixes to outlined in my pervious comment but everything seems to work.

It supports sharing text, images or any other files. Depending on the app you're sharing from you may be able to share multiple items at once (e.g. you can only share one note at a time from Google Keep but multiple images from gallery).

**Comment by roman-r-m** [comment]
One future improvement might be adding something similar to the desktop web clipper .
Right now when you share a page from e.g. Firefox (never tried other browsers but suspect they do the same) you get page title and URL.

Some note taking app like Google Keep just save it while Evernote for instance can fetch the content of the URL and save it as a note.

**Comment by CalebJohn** [comment]
Just letting you know that I am looking into this, I've tested on my phone sharing a variety of documents. The code will take more time so please bear with me.

File: ReactNativeClient/lib/components/select-date-time-dialog.js [hunk 2]
```diff
@@ -22,7 +22,7 @@ class SelectDateTimeDialog extends React.PureComponent {
 			this.setState({ date: newProps.date });
 		}
 
-		if ('shown' in newProps) {
+		if ('shown' in newProps && newProps.shown !== this.shown_) {
```
Comment (CalebJohn) [inline ReactNativeClient/lib/components/select-date-time-dialog.js:25]: Is this fix related to the share extension? Seems to be an unrelated bug fix.

File: ReactNativeClient/root.js [hunk 3]
```diff
@@ -660,6 +624,54 @@ class AppComponent extends React.Component {
...
+		const newNote = await Note.save({
+			parent_id: this.props.selectedFolderId,
+		}, { provisional: true });
+
+		setTimeout(() => {
+			this.props.dispatch({
+				type: 'NAV_GO',
+				routeName: 'Note',
+				noteId: newNote.id,
+				sharedData: {
```
Comment (CalebJohn) [inline ReactNativeClient/root.js:669]: Why isn't this just `sharedData: sharedData`? Is there some restriction on `ShareExtension.data()`?

Comment (roman-r-m) [reply ReactNativeClient/root.js:669]: Thanks, good catch. This piece went though quite a few changes and I lost track.

File: ReactNativeClient/lib/components/select-date-time-dialog.js (hunk 2 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/components/select-date-time-dialog.js:25]: Yes, it's not directly related.

It's just that with this change it started to occur every time (when sharing). And I spent a lot of time before realizing that it's been there all along and it's not that big of a deal. Though the warning does make it seem scary with the mention of memory leak.

But since I've fixed it anyway figured might as well include it (I'm a fan of the boy scout rule),

File: ReactNativeClient/root.js (hunk 3 above)
Comment (roman-r-m) [reply ReactNativeClient/root.js]: Fxied

**Comment by CalebJohn** [comment]
Alright This looks good to me! I think this is ready for a final review before getting merged. By the way, it looks like some tests are failing, but I think it's just because you don't have the latest changes from master. If you merge in master it should pass.

**Comment by roman-r-m** [comment]
Thanks, merged in master, let's see if this fixes tests.

**Comment by roman-r-m** [comment]
I think the test that fails is simply invalid. Also wrote here: https://discourse.joplinapp.org/t/broken-test-feature-forwardbackwardnotehistory/8282

File: ReactNativeClient/lib/components/screens/note.js [hunk 4]
```diff
@@ -121,6 +121,19 @@ class NoteScreenComponent extends BaseScreenComponent {
 				return true;
 			}
 
+			if (Platform.OS === 'android' && this.state.fromShare) {
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:124]: Can `state.fromShare` be set for iOS? If not, we can remove the platform check.

File: ReactNativeClient/lib/components/screens/note.js [hunk 5]
```diff
@@ -329,9 +342,9 @@ class NoteScreenComponent extends BaseScreenComponent {
 
 		shared.uninstallResourceHandling(this.refreshResource);
 
-		// if (Platform.OS !== 'ios' && this.state.fromShare) {
-		// 	ShareExtension.close();
-		// }
+		if (Platform.OS === 'android' && this.state.fromShare) {
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:345]: You've nicely abstracted away the platforms in lib/share.js you you can remove the platform check here.

File: ReactNativeClient/root.js [hunk 6]
```diff
@@ -660,6 +624,51 @@ class AppComponent extends React.Component {
 		});
 
 		AppState.addEventListener('change', this.onAppStateChange_);
+
+		const sharedData = Platform.OS === 'android' ? await ShareExtension.data() : null;
```
Comment (laurent22) [inline ReactNativeClient/root.js:628]: Here also the platform check is not needed.

File: ReactNativeClient/root.js [hunk 7]
```diff
@@ -660,6 +624,51 @@ class AppComponent extends React.Component {
...
+		if (sharedData) {
+			this.handleShared(sharedData);
+		}
+	}
+
+	async handleShared(sharedData) {
```
Comment (laurent22) [inline ReactNativeClient/root.js:634]: Could you move this to an external file, maybe something like root-handleShare.js? It would be a simple function that takes the required dependencies as input - sharedData, props.selectedFolderId, props.dispatch

Doing so would make it easier if some day we decide to refactor this huge root.js file.

File: ReactNativeClient/lib/components/screens/note.js [hunk 8]
```diff
@@ -523,7 +536,7 @@ class NoteScreenComponent extends BaseScreenComponent {
 
 		try {
 			if (mimeType == 'image/jpeg' || mimeType == 'image/jpg' || mimeType == 'image/png') {
-				const done = await this.resizeImage(localFilePath, targetPath, pickerResponse.mime);
+				const done = await this.resizeImage(localFilePath, targetPath, mimeType);
```
Comment (laurent22) [inline ReactNativeClient/lib/components/screens/note.js:539]: Why was this changed?

File: ReactNativeClient/lib/components/screens/note.js (hunk 4 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/components/screens/note.js:124]: Will do. Since I have no way of testing iOS I was maybe too defensive about trying not to break anything for iOS

File: ReactNativeClient/root.js (hunk 7 above)
Comment (roman-r-m) [reply ReactNativeClient/root.js:634]: Will do

File: ReactNativeClient/lib/components/screens/note.js (hunk 8 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/components/screens/note.js:539]: ImagePicker's response object does not have the mime field (see https://github.com/react-native-community/react-native-image-picker/blob/master/docs/Reference.md#the-response-object, I also verified in testing).
It's not a major issue, the only thing it affects is the extension of resized images.

File: ReactNativeClient/root.js (hunk 7 above)
Comment (roman-r-m) [reply ReactNativeClient/root.js]: This is done. I've extracted that function to a separate file (and rewrote in TypeScript). Also extracted away the part that handles permissions -- I plan to reuse it in another PR I have opened

**Comment by roman-r-m** [comment]
@laurent22 looks like I was wrong about PermissionsAndroid not working, tested again and seems ok this time. I'll remove the dependency on `react-native-permissions`

File: ReactNativeClient/root.js (hunk 6 above)
Comment (roman-r-m) [reply ReactNativeClient/root.js]: Rremoved

File: ReactNativeClient/lib/components/screens/note.js (hunk 5 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/components/screens/note.js]: Done

File: ReactNativeClient/lib/components/screens/note.js (hunk 4 above)
Comment (roman-r-m) [reply ReactNativeClient/lib/components/screens/note.js]: Done now

File: ReactNativeClient/lib/components/screens/note.js (hunk 8 above)
Comment (laurent22) [reply ReactNativeClient/lib/components/screens/note.js:539]: Good catch, that was indeed a bug!

File: ReactNativeClient/lib/permissions.ts [hunk 9]
```diff
@@ -0,0 +1,9 @@
+const { PermissionsAndroid } = require('react-native');
```
Comment (laurent22) [inline ReactNativeClient/lib/permissions.ts:1]: Please name the file "checkPermissions.ts" as this is how you import it below. In general the filename should the same as the exported item, to make searching in the code base easier. I'll see if there's an eslint rule for this to make it clearer.

File: ReactNativeClient/lib/share.js [hunk 10]
```diff
@@ -0,0 +1,13 @@
+const { NativeModules, Platform } = require('react-native');
```
Comment (laurent22) [inline ReactNativeClient/lib/share.js:1]: Please name the file ShareExtension.js, as you import it under this name (also could you convert it to TypeScript please?)

File: ReactNativeClient/lib/share-handler.ts [hunk 11]
```diff
@@ -0,0 +1,47 @@
+const Note = require('lib/models/Note.js');
```
Comment (laurent22) [inline ReactNativeClient/lib/share-handler.ts:1]: Could you name it "shareHandler.ts" please and import it under this name in root.js?

File: ReactNativeClient/root.js [hunk 12]
```diff
@@ -660,6 +624,16 @@ class AppComponent extends React.Component {
 		});
 
 		AppState.addEventListener('change', this.onAppStateChange_);
+
+		const sharedData = await ShareExtension.data();
+		if (sharedData) {
+			reg.logger().info(`Received shared data: ${JSON.stringify(sharedData)}`);
```
Comment (laurent22) [inline ReactNativeClient/root.js:630]: How large can the shared data be? Is there a chance stringifying a large object will crash the app? In any case, I think it would be safer to log minimal info here, for example just "Received shared data: title, body". Or even just "Received shared data".

Comment (roman-r-m) [reply ReactNativeClient/root.js]: Fixed

File: ReactNativeClient/lib/ShareExtension.ts [hunk 13]
```diff
@@ -10,4 +16,4 @@ const ext = (Platform.OS === 'android' && NativeModules.ShareExtension) ?
 		close: () => {},
 	};
 
-export default ext;
+export { ShareExtension };
```
Comment (laurent22) [inline ReactNativeClient/lib/ShareExtension.ts]: The `export default` was good actually. As a convention, all module should export a default member named like the filename (again I think there's an eslint for this, so I'll try to add this).

Comment (laurent22) [reply ReactNativeClient/lib/ShareExtension.ts]: Then you can import like so: `import ShareExtension, { SharedData } from './ShareExtension';`

Comment (roman-r-m) [reply ReactNativeClient/lib/ShareExtension.ts]: Done

**Comment by laurent22** [comment]
Ok we're good to go then, thanks a lot for this great feature, @roman-r-m!

**Comment by graphixillusion** [comment]
I just tried the new "share to joplin" feature under Android 9 and this is what i got:

1. Share to Joplin ---> Accetpt default name/location <---- WORKS
2. Share to Joplin ---> Change note location <---- DOESN'T WORK.

In option **#2**, Joplin just creates an empty note in the choosed destination named "**Untitled**"

**Comment by roman-r-m** [comment]
Raised a new issue to track this https://github.com/laurent22/joplin/issues/3343
